File: src/ai_guardrails/models/lint_issue.py
"""LintIssue model — lint finding with content-hash fingerprint."""

from __future__ import annotations

from dataclasses import dataclass
import hashlib


@dataclass(frozen=True)
class LintIssue:
    """A single lint finding. Line is display-only — matching uses fingerprint."""

    rule: str
    linter: str
    file: str
    line: int  # display only, NOT used for matching
    col: int
    message: str
    fingerprint: str  # computed from content context, not line number

    @staticmethod
    def compute_fingerprint(
        rule: str,
        file: str,
        line_content: str,
        context_before: list[str],
        context_after: list[str],
    ) -> str:
        """SHA-256 of rule + file + stripped context. Stable across line moves."""
        parts = [rule, file]
        parts += [line.strip() for line in context_before[-2:]]
        parts += [line_content.strip()]
        parts += [line.strip() for line in context_after[:2]]
        return hashlib.sha256("\n".join(parts).encode()).hexdigest()[:16]

File: src/models/lint-issue.ts
import { createHash } from "node:crypto";

export interface LintIssue {
  readonly rule: string; // "E501", "no-unused-vars", "S1481"
  readonly linter: string; // "ruff", "pyright", "biome"
  readonly file: string; // absolute path
  readonly line: number; // 1-indexed
  readonly col: number; // 1-indexed
  readonly message: string;
  readonly severity: "error" | "warning";
  readonly fingerprint: string; // content-stable SHA-256
}

export interface FingerprintOpts {
  rule: string;
  file: string;
  lineContent: string; // content of the flagged line
  contextBefore: string[]; // up to 2 lines before
  contextAfter: string[]; // up to 2 lines after
}

/**
 * Compute a fingerprint for a lint issue.
 * Hashes rule + file path + trimmed line content + surrounding context.
 * File path is included so identical issues in different files produce distinct fingerprints.
 *
 * NOTE: Callers must pass a project-relative path as `file`, NOT an absolute path.
 * Callers pass project-relative paths for portable baselines.
 */
export function computeFingerprint(opts: FingerprintOpts): string {
  const { rule, file, lineContent, contextBefore, contextAfter } = opts;
  const input = [
    rule,
    file,
    lineContent.trim(),
    ...contextBefore.map((l) => l.trim()),
    ...contextAfter.map((l) => l.trim()),
  ].join("\n");
  return createHash("sha256").update(input).digest("hex");
}
